### ISSUE TITLE
build failing on dokku

### ISSUE THREAD

**Issue by jomo** [crafatar/crafatar#95]
Not sure why this happens, it works on heroku.

```npm
Everything up-to-date
Counting objects: 26, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (25/25), done.
Writing objects: 100% (26/26), 2.72 KiB | 0 bytes/s, done.
Total 26 (delta 18), reused 0 (delta 0)
-----> Cleaning up ...
-----> Building crafatar ...
-----> Multipack app detected
       =====> Downloading Buildpack: https://github.com/mojodna/heroku-buildpack-cairo.git
       =====> Detected Framework: Cairo
-----> Vendoring binaries
       Fetching https://s3.amazonaws.com/mojodna-heroku/cedar-14/cairo-1.12.16-1.tar.gz
       Fetching https://s3.amazonaws.com/mojodna-heroku/cedar-14/freetype-2.5.3-1.tar.gz
       Fetching https://s3.amazonaws.com/mojodna-heroku/cedar-14/giflib-4.2.3-1.tar.gz
       Fetching https://s3.amazonaws.com/mojodna-heroku/cedar-14/pixman-0.32.6-1.tar.gz
-----> Tweaking Cairo, FreeType, and Pixman include paths
-----> Configuring build environment
-----> Building runtime environment
       =====> Downloading Buildpack: https://github.com/heroku/heroku-buildpack-nodejs.git
       =====> Detected Framework: Node.js

-----> Reading application state
       package.json...
       build directory...
       cache directory...
       environment variables...

       ACHIEVEMENT UNLOCKED: iojs :)

       Node engine:         1.2.x (iojs)
       Npm engine:          unspecified
       Start mechanism:     Procfile
       node_modules source: package.json
       node_modules cached: true

       NPM_CONFIG_PRODUCTION=true
       NODE_MODULES_CACHE=true

-----> Installing binaries
       Resolving iojs version 1.2.x via semver.io...
       Downloading and installing iojs 1.2.0...
       Using default npm version: 2.5.1

-----> Building dependencies
       Node version changed (v0.10.35 => v1.2.0); invalidating cache
       Installing node modules

       > canvas@1.2.1 install /tmp/build/node_modules/canvas
       > node-gyp rebuild

       child_process: customFds option is deprecated, use stdio instead.
       make: Entering directory `/tmp/build/node_modules/canvas/build'

       // stipped lots of output

       make: *** [Release/obj.target/canvas/src/CanvasRenderingContext2d.o] Error 1
       make: Leaving directory `/tmp/build/node_modules/canvas/build'
       gyp ERR! build error
       gyp ERR! stack Error: `make` failed with exit code: 2
       gyp ERR! stack     at ChildProcess.onExit (/tmp/build/.heroku/node/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
       gyp ERR! stack     at emitTwo (events.js:85:13)
       gyp ERR! stack     at ChildProcess.emit (events.js:153:7)
       gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1044:12)
       gyp ERR! System Linux 3.13.0-45-generic
       gyp ERR! command "node" "/tmp/build/.heroku/node/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
       gyp ERR! cwd /tmp/build/node_modules/canvas
       gyp ERR! node -v v1.2.0
       gyp ERR! node-gyp -v v1.0.2
       gyp ERR! not ok

       > lwip@0.0.6 install /tmp/build/node_modules/lwip
       > node-gyp rebuild

       child_process: customFds option is deprecated, use stdio instead.
       make: Entering directory `/tmp/build/node_modules/lwip/build'
 
       // stripped lots of output

       SOLINK_MODULE(target) Release/obj.target/lwip_image.node
       SOLINK_MODULE(target) Release/obj.target/lwip_image.node: Finished
       COPY Release/lwip_image.node
       make: Leaving directory `/tmp/build/node_modules/lwip/build'
       npm ERR! Linux 3.13.0-45-generic
       npm ERR! argv "/tmp/build/.heroku/node/bin/iojs" "/tmp/build/.heroku/node/bin/npm" "install" "--quiet" "--userconfig" "/tmp/build/.npmrc"
       npm ERR! node v1.2.0
       npm ERR! npm  v2.5.1
       npm ERR! code ELIFECYCLE

       npm ERR! canvas@1.2.1 install: `node-gyp rebuild`
       npm ERR! Exit status 1
       npm ERR!
       npm ERR! Failed at the canvas@1.2.1 install script 'node-gyp rebuild'.
       npm ERR! This is most likely a problem with the canvas package,
       npm ERR! not with npm itself.
       npm ERR! Tell the author that this fails on your system:
       npm ERR!     node-gyp rebuild
       npm ERR! You can get their info via:
       npm ERR!     npm owner ls canvas
       npm ERR! There is likely additional logging output above.

       npm ERR! Please include the following file with any support request:
       npm ERR!     /tmp/build/npm-debug.log

-----> Build failed

       We're sorry this build is failing! If you can't find the issue in application code,
       please submit a ticket so we can help: https://help.heroku.com/
       You can also try reverting to our legacy Node.js buildpack:
       heroku config:set BUILDPACK_URL=https://github.com/heroku/heroku-buildpack-nodejs#v63

       Love,
       Heroku
To dokku@crafatar:crafatar
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'dokku@crafatar:crafatar'
```

**Comment by josegonzalez** [comment]
Might be that the container is missing a C++ package to build the binary (I've had that same node-gyp issue on regular EC2 instances). Not sure what the trouble-shooting steps would be, perhaps trying to build the package yourself in a container and seeing what it needs?

**Comment by jomo** [comment]
I guess [this travis failure](https://travis-ci.org/crafatar/crafatar/jobs/51448899) is related. Apparently it's [caused by iojs / node-canvas](Automattic/node-canvas#520):
```man
> canvas@1.2.1 install /home/travis/build/crafatar/crafatar/node_modules/canvas
> node-gyp rebuild
child_process: customFds option is deprecated, use stdio instead.
make: Entering directory `/home/travis/build/crafatar/crafatar/node_modules/canvas/build'
  SOLINK_MODULE(target) Release/obj.target/canvas-postbuild.node
  SOLINK_MODULE(target) Release/obj.target/canvas-postbuild.node: Finished
  COPY Release/canvas-postbuild.node
  CXX(target) Release/obj.target/canvas/src/Canvas.o
  CXX(target) Release/obj.target/canvas/src/CanvasGradient.o
  CXX(target) Release/obj.target/canvas/src/CanvasPattern.o
  CXX(target) Release/obj.target/canvas/src/CanvasRenderingContext2d.o
../src/CanvasRenderingContext2d.cc: In member function ‘void Context2d::shadow(void (*)(cairo_t*))’:
../src/CanvasRenderingContext2d.cc:374:22: warning: unused variable ‘surface’ [-Wunused-variable]
../src/CanvasRenderingContext2d.cc: In static member function ‘static void Context2d::SetLineDash(const v8::FunctionCallbackInfo<v8::Value>&)’:
../src/CanvasRenderingContext2d.cc:2117:21: error: ‘_isnan’ was not declared in this scope
../src/CanvasRenderingContext2d.cc:2117:36: error: ‘_finite’ was not declared in this scope
../src/CanvasRenderingContext2d.cc: In static member function ‘static void Context2d::SetLineDashOffset(v8::Local<v8::String>, v8::Local<v8::Value>, const v8::PropertyCallbackInfo<void>&)’:
../src/CanvasRenderingContext2d.cc:2156:7: error: ‘_isnan’ was not declared in this scope
../src/CanvasRenderingContext2d.cc:2156:24: error: ‘_finite’ was not declared in this scope
make: *** [Release/obj.target/canvas/src/CanvasRenderingContext2d.o] Error 1
make: Leaving directory `/home/travis/build/crafatar/crafatar/node_modules/canvas/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/travis/.nvm/versions/io.js/v1.2.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at emitTwo (events.js:85:13)
gyp ERR! stack     at ChildProcess.emit (events.js:153:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1044:12)
gyp ERR! System Linux 2.6.32-042stab094.7
gyp ERR! command "node" "/home/travis/.nvm/versions/io.js/v1.2.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/travis/build/crafatar/crafatar/node_modules/canvas
gyp ERR! node -v v1.2.0
gyp ERR! node-gyp -v v1.0.2
gyp ERR! not ok 
```

**Comment by jomo** [comment]
It was indeed an issue with canvas & iojs, fixed by Automattic/node-canvas#522
